File: src/execution/runtimeSimulationEngine.ts
/**
 * Converted from Python: core/execution/runtime_simulation_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { executeRuntimeAction } from "./runtimeExecutionEngine.js";
import { trackRuntimeMutations } from "./runtimeMutationEngine.js";
import { buildRuntimeSandbox } from "./runtimeSandboxEngine.js";

export function simulateRuntimeExecution(actions: any, sandbox: any = null, tick: any = 0): any {
  sandbox = py.or2(sandbox, () => (buildRuntimeSandbox()));
  var predicted: any[] = [];
  var rollback_required: any = false;
  var index: any;
  var raw: any;
  for ([index, raw] of py.enumerate(py.slice(actions, null, 1000))) {
    var result: any = executeRuntimeAction(raw, sandbox, undefined, undefined, py.add(tick, index));
    if (py.truthy(py.get(result, "executed"))) {
      py.listAppend(predicted, {"kind": py.toStr(py.get(raw, "type", "action")), "target": py.toStr(py.get(raw, "selector", py.get(raw, "window", py.get(raw, "command", "")))), "tick": py.add(tick, index)});
    } else {
      rollback_required = true;
    }
  }
  var mutation_view: any = trackRuntimeMutations(undefined, {"kind": "simulated", "target": "dry_run", "tick": tick});
  return {"simulated": true, "predicted_mutations": predicted, "rollback_required": rollback_required, "mutation_preview": py.get(mutation_view, "mutations", []), "runtime_mutated": false, "bounded": true};
}
export { buildRuntimeSandbox, executeRuntimeAction, trackRuntimeMutations };

File: lib/src/execution/runtime_simulation_engine.dart
import 'runtime_execution_engine.dart';
import 'runtime_mutation_engine.dart';
import 'runtime_sandbox_engine.dart';

/// Port of core/execution/runtime_simulation_engine.simulate_runtime_execution.
Map<String, dynamic> simulateRuntimeExecution(
  List<Map<String, dynamic>> actions, {
  Map<String, dynamic>? sandbox,
  int tick = 0,
}) {
  final Map<String, dynamic> sb = sandbox ?? buildRuntimeSandbox();
  final List<Map<String, dynamic>> predicted = <Map<String, dynamic>>[];
  bool rollbackRequired = false;

  final List<Map<String, dynamic>> bounded =
      actions.length > 1000 ? actions.sublist(0, 1000) : actions;

  for (int index = 0; index < bounded.length; index++) {
    final Map<String, dynamic> raw = bounded[index];
    final Map<String, dynamic> result =
        executeRuntimeAction(raw, sandbox: sb, tick: tick + index);
    if (result['executed'] == true) {
      // Python: raw.get("selector", raw.get("window", raw.get("command", "")))
      final dynamic target = raw.containsKey('selector')
          ? raw['selector']
          : raw.containsKey('window')
              ? raw['window']
              : raw.containsKey('command')
                  ? raw['command']
                  : '';
      predicted.add(<String, dynamic>{
        'kind': '${raw['type'] ?? 'action'}',
        'target': '$target',
        'tick': tick + index,
      });
    } else {
      rollbackRequired = true;
    }
  }

  final Map<String, dynamic> mutationView = trackRuntimeMutations(
    mutation: <String, dynamic>{
      'kind': 'simulated',
      'target': 'dry_run',
      'tick': tick,
    },
  );

  return <String, dynamic>{
    'simulated': true,
    'predicted_mutations': predicted,
    'rollback_required': rollbackRequired,
    'mutation_preview': mutationView['mutations'],
    'runtime_mutated': false,
    'bounded': true,
  };
}
